### PR TITLE
add salt to password for sha256_password users

### PR DESCRIPTION
**Pull request by varadarajkumar**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
...

> Information about CI checks: https://clickhouse.tech/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by alexey-milovidov** [comment]
Wrong diff, needs rebasing.

File: src/Access/AccessEntityIO.cpp [hunk 1]
```diff
@@ -85,6 +85,16 @@ String serializeAccessEntity(const IAccessEntity & entity)
     WriteBufferFromOwnString buf;
     for (const ASTPtr & query : queries)
     {
+        if (const User * user = typeid_cast<const User *>(&entity))
+        {
+            if (user->auth_data.getSalt().compare("") != 0)
+            {
+                std::string strSalt = " '";
+                strSalt += user->auth_data.getSalt();
+                strSalt+="'";
```
Comment (alexey-milovidov) [inline src/Access/AccessEntityIO.cpp:94]: Wrong style.

File: src/Access/AccessEntityIO.cpp [hunk 2]
```diff
@@ -85,6 +85,16 @@ String serializeAccessEntity(const IAccessEntity & entity)
     WriteBufferFromOwnString buf;
     for (const ASTPtr & query : queries)
     {
+        if (const User * user = typeid_cast<const User *>(&entity))
+        {
+            if (user->auth_data.getSalt().compare("") != 0)
+            {
+                std::string strSalt = " '";
```
Comment (alexey-milovidov) [inline src/Access/AccessEntityIO.cpp:92]: Why quotes?

Comment (varadarajkumar) [reply src/Access/AccessEntityIO.cpp:92]: Hi Alexey, 
it was a wrong code base. I will close this and create new PR. sorry about this.

**Comment by varadarajkumar** [comment]
wrongly created on master.

File: src/Access/AccessEntityIO.cpp [hunk 3]
```diff
@@ -85,6 +85,16 @@ String serializeAccessEntity(const IAccessEntity & entity)
     WriteBufferFromOwnString buf;
     for (const ASTPtr & query : queries)
     {
+        if (const User * user = typeid_cast<const User *>(&entity))
+        {
+            if (user->auth_data.getSalt().compare("") != 0)
```
Comment (alexey-milovidov) [inline src/Access/AccessEntityIO.cpp:90]: empty()

File: src/Access/Authentication.cpp [hunk 4]
```diff
@@ -31,9 +31,9 @@ namespace
         return (Util::encodeDoubleSHA1(password) == password_double_sha1);
     }
 
-    bool checkPasswordSHA256(const std::string_view & password, const Digest & password_sha256)
+    bool checkPasswordSHA256(const std::string_view & password, const Digest & password_sha256, const String & salt)
```
Comment (alexey-milovidov) [inline src/Access/Authentication.cpp:34]: string_view vs String

File: src/Access/Authentication.cpp [hunk 5]
```diff
@@ -31,9 +31,9 @@ namespace
...
     {
-        return Util::encodeSHA256(password) == password_sha256;
+        return Util::encodeSHA256(String(password)+salt) == password_sha256;
```
Comment (alexey-milovidov) [inline src/Access/Authentication.cpp:36]: Concatenation is not always ok, see "length extension attack" and HMAC.

File: src/Access/Common/AuthenticationData.h [hunk 6]
```diff
@@ -54,7 +54,7 @@ class AuthenticationData
 public:
     using Digest = std::vector<uint8_t>;
 
-    explicit AuthenticationData(AuthenticationType type_ = AuthenticationType::NO_PASSWORD) : type(type_) {}
+    explicit AuthenticationData(AuthenticationType type_ = AuthenticationType::NO_PASSWORD) : type(type_), salt("") {}
```
Comment (alexey-milovidov) [inline src/Access/Common/AuthenticationData.h:57]: It is default initialized to empty string.

File: src/Access/Common/AuthenticationData.h [hunk 7]
```diff
@@ -76,6 +76,10 @@ class AuthenticationData
     void setPasswordHashBinary(const Digest & hash);
     const Digest & getPasswordHashBinary() const { return password_hash; }
 
+    /// Sets the salt in String form.
+    void setSalt(const String & salt);
```
Comment (alexey-milovidov) [inline src/Access/Common/AuthenticationData.h:80]: Passing by const reference here is suboptimal, better pass by value and move.

File: src/Parsers/Access/ParserCreateUserQuery.cpp [hunk 8]
```diff
@@ -141,6 +144,16 @@ namespace
             }
 
             auth_data = AuthenticationData{*type};
+
+            if (type == AuthenticationType::SHA256_PASSWORD)
```
Comment (alexey-milovidov) [inline src/Parsers/Access/ParserCreateUserQuery.cpp:148]: Need to make an integration test including server restart.

File: src/Access/AccessEntityIO.cpp [hunk 9]
```diff
@@ -85,6 +85,16 @@ String serializeAccessEntity(const IAccessEntity & entity)
...
+            {
+                std::string strSalt = " '";
+                strSalt += user->auth_data.getSalt();
```
Comment (alexey-milovidov) [inline src/Access/AccessEntityIO.cpp:93]: The string is not escaped.

File: src/Access/AccessEntityIO.cpp [hunk 10]
```diff
@@ -121,6 +131,9 @@ AccessEntityPtr deserializeAccessEntityImpl(const String & definition)
                 throw Exception("Two access entities attached in the same file", ErrorCodes::INCORRECT_ACCESS_ENTITY_DEFINITION);
             res = user = std::make_unique<User>();
             InterpreterCreateUserQuery::updateUserFromQuery(*user, *create_user_query);
+
+            if (user->auth_data.getSalt().compare("") != 0)
```
Comment (alexey-milovidov) [inline src/Access/AccessEntityIO.cpp:135]: empty()

Comment (varadarajkumar) [reply src/Access/AccessEntityIO.cpp:135]: sure i will update it to empty()

File: src/Access/AccessEntityIO.cpp (hunk 9 above)
Comment (varadarajkumar) [reply src/Access/AccessEntityIO.cpp:93]: the random salt saving in <UUID>.sql file within quotes, after password hash. which would be appended during user login scenario to the input password.

**Pull request closed without merging**